### PR TITLE
Apply vertical exaggeration to box-shaped voxels

### PR DESCRIPTION
**Pull request by jjhembd**
# Description

This PR applies vertical exaggeration to BOX-shaped voxels. The voxel is assumed to be oriented with its z-axis along the local ellipsoid normal.

The exaggeration is applied by applying a scale and shift to the model transforms. The transforms for voxels are applied in the following sequence:

1. Shape transform. This scales the box to the appropriate size, and accounts for any shift in the center
2. Model matrix. This is the identity by default, but can be modified by the end user.
3. _New_: Exaggeration transform.
4. Root tile transform. This moves the box to the appropriate position on the globe.

Since the new exaggeration transform is applied before the root tile transform, the local Z coordinate is assumed to be along the "up" direction. However, the root tile transform is included in the calculation of the shift needed to honor `Scene.verticalExaggerationRelativeHeight`, by transforming the BOX center (from the shape transform) and computing its geodetic height.

image

image

## Testing plan

Load [this local Sandcastle](http://localhost:8080/Apps/Sandcastle/index.html#c=tVh7U9u4Fv8qHv5Z0wYnEOjcksBcSmGWGShMQzvdls6uYiuJLrLkK8lpzA7ffY9etuWElu7em8mMpaPz/Ono6JFyJlW0JPgbFtFRxPC36BRLUubJR0OL77ZS0z/lTCHCsLjb6kV/3rEomiKJL1GFxaEXucjRHIvKEJOZ4PmN4EuSYXEiK5bGWijyvLeE4itUTLBYkhQ7Sc9vhD8I6kRqoWlJaHbFs5JiM7p1IiVWsn+LV6oUWPbfIfgieoaEWlxc3G1tWwXms90LvL4h6b32fYaoxGZojnnKs5CmSI4phN2mIUZypAhnDfFxe3TH7pjFMQFvMMti57QlfuQrTC+YLHCquLgiK8KsTGpm4M9IppjhXpSiHAsUPcJkWEFgMkNJhqflfLLg384F8MgbLCYYhDPgVKLENV/BpQIgYdrkRAGsMpmtEEowQ1OKW9zetDYDmGIKQ2Zi8QrNYTZciFbpEgtFUkC2NWbAEJhCZ4l/xWS+UN9jfx9watCME7OSpXo4KosMKdyWiLetQ0/rBJfflfkU8rSOImm7ryGOnu1To82mTaMzjNJkU0uzwP8tsVTvYdJBVo886sjc/N8znt7zUiVKoPS+cVTz2RlQnNMp0usv42mZY6aSOVZnFOvmm+oig0x3PJDSo3XNqCho9YawjLC5bCz0vGYtM+Miiq09BvkTEdbE53Ams6iF4wLJ628MMqkA4KpYC207xnpBeg/8OtV+X08lLGqdbG1PjHjNJsupTAWZ4nh91i2yjw5E6zEwI5UuTjnlnTJlSPF2mEom/TNdCa5KqoiuNWb9+foSywUqsAtGLYhMDAE0m++oJueEveElyyQMQWhXvusUtBjRKmD03TXGDAoKk+Ck7MQBJQtaiA3jf/Ui+O+3hDR4mv8LVGMd793W12ZQVYUddJqusEKAKboFevLx7HS/xZvyvOAMsur2CaHTNkNyfnl9cjvcayn4neIlpqcQm14tQ5/qNfIbQGpyy4F8dFTvMXpSJppqzJ1dXl7cTK4v3tZpZme/PQtrcCUpBYe7O0VH8wa33M4wst/aRLICI4NksEavgL5zhdQiubmI+tHBBpYHzbI3ML96WGDYkljD5bPbFk8zBmQ8g00m2whnJ5X+B3C2cvXvw9lxqwOnHzVwNqjtNajVHFXAcbCBQ+P6FKye6Tmw/qgsJIXgiuv15Ev6W1gSYLyeD4MLL3Rb3jGHq9/BFei71MujZ5qf7Oc3+/lstnQnOrJ+6kmshaLjo7UlVk/chojqaK35oK50Ks2oYdPBeQZTOL4MvraGBWIZzydYnxOsZev7i5Z+mLCgu4pe2kA3kz+11JuyBDTwwBBEmar3xqSmarDjRkPPONtr+bQ9CuYXpi0nUs+V5HSJ4y9e/VezBYdbwt+yZ+H3JVLnKBw3rQJ7UNBcceigj3WpM8sXygCX74H50NKQLhBjjY5N5R0WYl2xDV+sY9juQN7eNs8pR2q4dyIEquKWjy8Ccx5pc2igWEV6CQ5G8BmH3kYPL1/WWVpzV5a7CrkroLS4W/wry78K+VdACfh9TBQLOJN8AqEVVIw4AHcn2vVFqCvwGwhUHYHqewKfQeChI/DgBRqRft8KLYB9UWLIe+/gC9hJDkDA29f9YT3+2fT31owvWpOtc47BfaKddPG6v3rR7yaDjql11dRsbgdr9NRlSPtQZS5hv0oaL3oRLA6oamChFxzFQhysKjiF4pWvHvb38MP6sYG0WnPSrrO9n0cH0WKBLI/fa2aUw7HRa3ypQQmDqdfNFxtQuD60xFdTxTRQwhXkZ0ruNpJzgTH7Gdm9RnZKS/wzokMt2oKiln20jcf1HbTWOQpO42kpFRS8Bcq6jwanrZHYrNuZQHN9jbE0fVE/jP5YcpLVI1eIsPjcdS5YUapoJs23B8kEl4sofch/z/U1AqYOC4IobPu24Y4ddYVY4nRoK+87LnKk052ZBnnAsVZjewCOM5EYZriPiBlKsRVqEmimiyUk5mxWSr1pwWkjHuhFkHEVt8z0jItU3w3fEoHNjnN2ut1VZBjgimYXoc06k/hGfyv9fHhJY9r7m7vyn7jUm09Bg1c8WlNg5vtpceQkHv9oHjHcK4IgOdFXXgnpnfMlPqE0br9YSH8ajJ46f765/lRfcQt3xHLp8rwL2q3by2olXBJ362+dW/lcoGJB0kTx+hhrzmmd21XNp+vaW6yXnox3dneHyd6rVwcHw/1eNHyVDF4PXw/3D6DYDQaQXc0lfcpXkxRRvPHQrFW6+7qM981JFW5w7muPrlqTR8FeN2+h/EjYA/Ogogmy2jf6jLXYmw3E55RPEd0gX5NkgpFUkJ1q8aG45edkhTPzbhR7EJvIlhZ5N9/6EtzNAJRlXUQ/BkKxW4Hew8O61XM3j1ZhOAx6huHRQR26ArUdweFOTVBeULtu3NtVh+8/RCmTW+Y9rv0Utwye3ZLmdWMt6JCgwTEvcbBTVLfc3DHAg0mxwALHHfPTYNQ9j+rsti9oumK41bXV2xpLVVF8rFn+TeDoJlRUChonSV9hCBMWruxPy/QeqySVUguN+15knJFlRLKjDY+yUUqRlDAyKymdQLm72zoe94E/EIMypB29XmJBUaVZFrvHl5aYJMm4D911qfrtyXg9Vvpt59jO61hNeVYd+7ozVuK42ZLGKjtuv+yM+0AIh9vHhDExxV+fYMEoZPIcgtAXZ+jt6hZaQesAWlLhApoArKbrerYzhV0PSEsEu+Jh8ITZiwzxg3loOox+MVZ+ccFstq1gm9JmAEZn8Qc2Am1BmNAR34HHP0BG9mXxHyC0s6urTA3Ta9dzUD2BU/iu+f9CqvN6+hysoFknFrRdxvnU/As) and verify the procedural BOX voxel exaggerates as expected.

# Author checklist

- [x] I have submitted a Contributor License Agreement
- [x] I have added my name to `CONTRIBUTORS.md`
- [x] I have updated `CHANGES.md` with a short summary of my change
- [ ] I have added or updated unit tests to ensure consistent code coverage
- [x] I have update the inline documentation, and included code examples where relevant
- [x] I have performed a self-review of my code

### CODE REVIEW

**Comment by ggetz** [comment]
@jjhembd Please merge in `main` when you get the chance.

**Comment by jjhembd** [comment]
Thanks @ggetz, I added a [local Sandcastle](http://localhost:8080/Apps/Sandcastle/index.html#c=tVh7U9u4Fv8qHv5Z0wYnEOjcksBcSmGWGShMQzvdls6uYiuJLrLkK8lpzA7ffY9etuWElu7em8mMpaPz/Ono6JFyJlW0JPgbFtFRxPC36BRLUubJR0OL77ZS0z/lTCHCsLjb6kV/3rEomiKJL1GFxaEXucjRHIvKEJOZ4PmN4EuSYXEiK5bGWijyvLeE4itUTLBYkhQ7Sc9vhD8I6kRqoWlJaHbFs5JiM7p1IiVWsn+LV6oUWPbfIfgieoaEWlxc3G1tWwXms90LvL4h6b32fYaoxGZojnnKs5CmSI4phN2mIUZypAhnDfFxe3TH7pjFMQFvMMti57QlfuQrTC+YLHCquLgiK8KsTGpm4M9IppjhXpSiHAsUPcJkWEFgMkNJhqflfLLg384F8MgbLCYYhDPgVKLENV/BpQIgYdrkRAGsMpmtEEowQ1OKW9zetDYDmGIKQ2Zi8QrNYTZciFbpEgtFUkC2NWbAEJhCZ4l/xWS+UN9jfx9watCME7OSpXo4KosMKdyWiLetQ0/rBJfflfkU8rSOImm7ryGOnu1To82mTaMzjNJkU0uzwP8tsVTvYdJBVo886sjc/N8znt7zUiVKoPS+cVTz2RlQnNMp0usv42mZY6aSOVZnFOvmm+oig0x3PJDSo3XNqCho9YawjLC5bCz0vGYtM+Miiq09BvkTEdbE53Ams6iF4wLJ628MMqkA4KpYC207xnpBeg/8OtV+X08lLGqdbG1PjHjNJsupTAWZ4nh91i2yjw5E6zEwI5UuTjnlnTJlSPF2mEom/TNdCa5KqoiuNWb9+foSywUqsAtGLYhMDAE0m++oJueEveElyyQMQWhXvusUtBjRKmD03TXGDAoKk+Ck7MQBJQtaiA3jf/Ui+O+3hDR4mv8LVGMd793W12ZQVYUddJqusEKAKboFevLx7HS/xZvyvOAMsur2CaHTNkNyfnl9cjvcayn4neIlpqcQm14tQ5/qNfIbQGpyy4F8dFTvMXpSJppqzJ1dXl7cTK4v3tZpZme/PQtrcCUpBYe7O0VH8wa33M4wst/aRLICI4NksEavgL5zhdQiubmI+tHBBpYHzbI3ML96WGDYkljD5bPbFk8zBmQ8g00m2whnJ5X+B3C2cvXvw9lxqwOnHzVwNqjtNajVHFXAcbCBQ+P6FKye6Tmw/qgsJIXgiuv15Ev6W1gSYLyeD4MLL3Rb3jGHq9/BFei71MujZ5qf7Oc3+/lstnQnOrJ+6kmshaLjo7UlVk/chojqaK35oK50Ks2oYdPBeQZTOL4MvraGBWIZzydYnxOsZev7i5Z+mLCgu4pe2kA3kz+11JuyBDTwwBBEmar3xqSmarDjRkPPONtr+bQ9CuYXpi0nUs+V5HSJ4y9e/VezBYdbwt+yZ+H3JVLnKBw3rQJ7UNBcceigj3WpM8sXygCX74H50NKQLhBjjY5N5R0WYl2xDV+sY9juQN7eNs8pR2q4dyIEquKWjy8Ccx5pc2igWEV6CQ5G8BmH3kYPL1/WWVpzV5a7CrkroLS4W/wry78K+VdACfh9TBQLOJN8AqEVVIw4AHcn2vVFqCvwGwhUHYHqewKfQeChI/DgBRqRft8KLYB9UWLIe+/gC9hJDkDA29f9YT3+2fT31owvWpOtc47BfaKddPG6v3rR7yaDjql11dRsbgdr9NRlSPtQZS5hv0oaL3oRLA6oamChFxzFQhysKjiF4pWvHvb38MP6sYG0WnPSrrO9n0cH0WKBLI/fa2aUw7HRa3ypQQmDqdfNFxtQuD60xFdTxTRQwhXkZ0ruNpJzgTH7Gdm9RnZKS/wzokMt2oKiln20jcf1HbTWOQpO42kpFRS8Bcq6jwanrZHYrNuZQHN9jbE0fVE/jP5YcpLVI1eIsPjcdS5YUapoJs23B8kEl4sofch/z/U1AqYOC4IobPu24Y4ddYVY4nRoK+87LnKk052ZBnnAsVZjewCOM5EYZriPiBlKsRVqEmimiyUk5mxWSr1pwWkjHuhFkHEVt8z0jItU3w3fEoHNjnN2ut1VZBjgimYXoc06k/hGfyv9fHhJY9r7m7vyn7jUm09Bg1c8WlNg5vtpceQkHv9oHjHcK4IgOdFXXgnpnfMlPqE0br9YSH8ajJ46f765/lRfcQt3xHLp8rwL2q3by2olXBJ362+dW/lcoGJB0kTx+hhrzmmd21XNp+vaW6yXnox3dneHyd6rVwcHw/1eNHyVDF4PXw/3D6DYDQaQXc0lfcpXkxRRvPHQrFW6+7qM981JFW5w7muPrlqTR8FeN2+h/EjYA/Ogogmy2jf6jLXYmw3E55RPEd0gX5NkgpFUkJ1q8aG45edkhTPzbhR7EJvIlhZ5N9/6EtzNAJRlXUQ/BkKxW4Hew8O61XM3j1ZhOAx6huHRQR26ArUdweFOTVBeULtu3NtVh+8/RCmTW+Y9rv0Utwye3ZLmdWMt6JCgwTEvcbBTVLfc3DHAg0mxwALHHfPTYNQ9j+rsti9oumK41bXV2xpLVVF8rFn+TeDoJlRUChonSV9hCBMWruxPy/QeqySVUguN+15knJFlRLKjDY+yUUqRlDAyKymdQLm72zoe94E/EIMypB29XmJBUaVZFrvHl5aYJMm4D911qfrtyXg9Vvpt59jO61hNeVYd+7ozVuK42ZLGKjtuv+yM+0AIh9vHhDExxV+fYMEoZPIcgtAXZ+jt6hZaQesAWlLhApoArKbrerYzhV0PSEsEu+Jh8ITZiwzxg3loOox+MVZ+ccFstq1gm9JmAEZn8Qc2Am1BmNAR34HHP0BG9mXxHyC0s6urTA3Ta9dzUD2BU/iu+f9CqvN6+hysoFknFrRdxvnU/As) for testing.

File: packages/engine/Source/Scene/VoxelPrimitive.js [hunk 1]
```diff
@@ -1154,14 +1170,91 @@ function updateVerticalExaggeration(primitive, frameState) {
     primitive._maxBounds,
     primitive._exaggeratedMaxBounds
   );
-  if (defined(frameState) && primitive.shape === VoxelShapeType.ELLIPSOID) {
+  if (!defined(frameState)) {
```
Comment (ggetz) [inline packages/engine/Source/Scene/VoxelPrimitive.js:1173]: Why is `frameState` optional? Even if some shapes don't use it, we can still pass it in all the time, unless there is some other part of the architecture that is prevent that.

**Comment by ggetz** [comment]
Thanks @jjhembd! I've testing this out and the behavior seems to be working well! I juts have that one comment on general code structure.

Is it reasonable to add a test here?

File: packages/engine/Source/Scene/VoxelPrimitive.js (hunk 1 above)
Comment (jjhembd) [reply packages/engine/Source/Scene/VoxelPrimitive.js]: I was calling `updateVerticalExaggeration` from the constructor (via `initialize`), when the `frameState` was not available. But this does create unnecessary branching.

I updated the `initialize` method to directly set the initial values for the exaggerated bounds & matrix, so `updateVerticalExaggeration` always follows the same path now.

**Comment by ggetz** [comment]
Thanks @jjhembd! Should a test be added here, or is this ready to go?

**Comment by jjhembd** [comment]
Hi @ggetz, I added two exaggeration-related tests to VoxelPrimitiveSpec, to test exaggeration of both ellipsoid- and box-shaped voxels.

**Comment by ggetz** [comment]
Fantastic thanks @jjhembd!

**Pull request merged**